dotnet/src/Agents/Abstractions/ChatHistoryChannel.cs

```diff
@@ -25,16 +26,55 @@ public class ChatHistoryChannel : AgentChannel
             throw new KernelException($"Invalid channel binding for agent: {agent.Id} ({agent.GetType().FullName})");
         }
 
-        await foreach (ChatMessageContent message in historyHandler.InvokeAsync(this._history, cancellationToken).ConfigureAwait(false))
+        // Capture the current message count to evaluate history mutation.
+        int messageCount = this._history.Count;
+        HashSet<ChatMessageContent> mutatedHistory = [];
+
+        // Utilize a queue as a "read-ahead" cache to evaluate message sequencing (i.e., which message is final).
+        Queue<ChatMessageContent> messageQueue = [];
+
+        ChatMessageContent? yieldMessage = null;
+        await foreach (ChatMessageContent responseMessage in historyHandler.InvokeAsync(this._history, cancellationToken).ConfigureAwait(false))
```

Comment:
Agree.  I believe we have resounding consensus based on our group coversation that this is an issue.  Current thinking is that this may be addressed as part of supporting a history truncation strategy:

- https://github.com/microsoft/semantic-kernel/issues/7320
- https://github.com/microsoft/semantic-kernel/issues/7321